core/src/test/java/org/apache/calcite/test/RelMetadataTest.java

```diff
@@ -1618,20 +1619,23 @@ private void checkCollation(RelOptCluster cluster, RelOptTable empTable,
     final EnumerableMergeJoin join;
     join = EnumerableMergeJoin.create(project, deptSort,
         rexBuilder.makeLiteral(true), leftKeys, rightKeys, JoinRelType.INNER);
+    assertThat(join.getTraitSet().getConvention(), equalTo(EnumerableConvention.INSTANCE));
```

Comment:
Should this check be moved to `EnumerableMergeJoin` constructor?

Reply:
Other similar operators (e.g. `EnumerableHashJoin`, `EnumerableNestedLoopJoin`, `EnumerableBatchNestedLoopJoin`) don't have this check in their constructors. In fact, it is not recommended to call these constructors directly (unless you know what you are doing) and they are not public, so I am ok with their current behavior of "trusting" the `RelTraitSet` parameter.
On the contrary, `EnumerableMergeJoin#create` method is public (as the other `create` methods in the other join operators), so it is its responsibility to create a valid `traitSet` bafore calling `EnumerableMergeJoin`'s constructor.

Reply:
See https://github.com/apache/calcite/blob/978bb7ea44969351468d1b5e240e8f57af7e5770/core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCalc.java#L77 and other assertions of that kind: https://github.com/apache/calcite/search?q=%22getConvention%28%29+instanceof+EnumerableConvention%22

Reply:
Ok, I'm not against having the assertion in the constructor. I'll add it.

Reply:
Great. Then the assertion in the test won't be needed

Reply:
I'd prefer to keep it in the unit test, so that we can detect any potential regression in the future.

Reply:
Assertion added to `EnumerableMergeJoin`'s constructor.

Reply:
I think we do not need to duplicate logic though.
For instance, if the method receives `String`, then we don't need to add a test that tries to pass `Integer` there.

The assertion in the constructor covers all the possible ways to create `EnumerableMergeJoin`, so it is more-or-less enough from my point of view.

Reply:
Ok, I don't mind removing the check from the test. I just want this fix merged before next release.

Reply:
Done.